boards: Add initial support for Adafruit Macropad RP2040

kartben opened the pull request:
https://builds.zephyrproject.io/zephyr/pr/85494/docs/boards/adafruit/macropad_rp2040/doc/index.html

The board is based on the Raspberry Pi RP2040 microcontroller and includes a 3x4 mechanical keyboard, 12 RGB NeoPixels, a rotary encoder with push button, a 128x64 OLED display, and a small speaker for audio feedback.

Tested:

- [x] UART on stemma port
- [x] LED
- [x] LED Strip (12 pixels)
- [x] PWM buzzer (and associated regulator)
- [x] Keypad (12 keys)
- [x] LCD display (SH1106)
- [x] USB device (console, MIDI, ...)
- [x] Clickable rotary encoder

![5128-10](https://github.com/user-attachments/assets/cd6e2cd4-2495-4292-8b77-673d403581bf)

kartben committed:
boards: adafruit: add initial support for Macropad RP2040

This commit adds initial support for the Adafruit MacroPad RP2040.

The board is based on the Raspberry Pi RP2040 microcontroller and
includes a 3x4 mechanical keyboard, 12 RGB NeoPixels, a rotary encoder
with push button, a 128x64 OLED display, and a small speaker for audio
feedback.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

fabiobaltieri commented on boards/adafruit/macropad_rp2040/adafruit_macropad_rp2040.dts:
```diff
@@ -0,0 +1,256 @@
+			gpios = <&gpio0 19 GPIO_ACTIVE_HIGH>;
+			chain-length = <12>;
+			color-mapping = <LED_COLOR_ID_GREEN
+			LED_COLOR_ID_RED
+			LED_COLOR_ID_BLUE>;
```
indentation is off

kartben replied:
done thx!

fabiobaltieri commented on boards/adafruit/macropad_rp2040/adafruit_macropad_rp2040-pinctrl.dtsi:
```diff
@@ -0,0 +1,50 @@
+			pinmux = <PWM_0A_P16>;
+		};
+	};
+
```
trailing-ish space, drop

kartben replied:
woops - thx

fabiobaltieri commented on boards/adafruit/macropad_rp2040/adafruit_macropad_rp2040.dts:
```diff
@@ -0,0 +1,256 @@
+
+		key0: key_0 {
+			gpios = <&gpio0 1 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>;
+			label = "Key 0";
```
more of a suggestion but I usually avoid these labels, they are not used anywhere, they don't add any more information from what you already have from the node name ("key_0") and it just become one more thing you are going to forget to rename after copy pasting it 9 times, I suggest dropping these from all keys and leds

kartben replied:
ya, makes sense. I've dropped from keys, left it for the LED as it seems to be common practice for most of the Pico boards (and beyond, really, for better or worse). Thanks for the quick review!

ajf58 commented on boards/adafruit/macropad_rp2040/adafruit_macropad_rp2040.yaml:
```diff
@@ -0,0 +1,23 @@
+ram: 264
+flash: 8192
+supported:
+  - uart
+  - gpio
+  - i2c
+  - spi
+  - pwm
+  - usb_device
+  - display
+  - qdec
+  - led_strip
```
```suggestion
  - display
  - gpio
  - i2c
  - led_strip
  - pwm
  - qdec
  - spi
  - uart
  - usb_device
```
Non-blocking: sort alphabetically.

soburi replied:
In the current situation, I think it would be better to align it with rpi_pico from the viewpoint of ease of maintenance. If you want to fix it, it would be better to do it all at once with the rp2040 board.

ajf58 commented on boards/adafruit/macropad_rp2040/adafruit_macropad_rp2040.yaml:
```diff
@@ -0,0 +1,23 @@
+  - display
+  - qdec
+  - led_strip
+testing:
+  ignore_tags:
+    - net
+    - bluetooth
```
Are these required? I can't see any connectivity options (BLE, Wi-Fi, cellular) on the board.
```suggestion
```

kartben replied:
ya, of course! thanks - dropped

ajf58 commented on boards/adafruit/macropad_rp2040/adafruit_macropad_rp2040_defconfig:
```diff
@@ -0,0 +1,13 @@
+CONFIG_BUILD_OUTPUT_UF2=y
+CONFIG_UART_INTERRUPT_DRIVEN=y
+CONFIG_RESET=y
+CONFIG_CLOCK_CONTROL=y
```
This is a fairly feature-rich board. Should more functionality be enabled by default?

ajf58 commented on boards/adafruit/macropad_rp2040/board.cmake:
```diff
@@ -0,0 +1,36 @@
+# And any other SWD debug adapter might also be usable with this configuration.
+
+# Set RPI_PICO_DEBUG_ADAPTER to select debug adapter by command-line arguments.
+# e.g.) west build -b rpi_pico -- -DRPI_PICO_DEBUG_ADAPTER=raspberrypi-swd
```
```suggestion
# e.g. west build -b adafruit_macropad_rp2040 -- -DRPI_PICO_DEBUG_ADAPTER=raspberrypi-swd
```
Side-note: we need a better way to include this into `board.cmake` files. A lot of this is duplicated for every RP2040-based board.

kartben replied:
dropped altogether to only keep uf2

ajf58 commented on boards/adafruit/macropad_rp2040/board.cmake:
```diff
@@ -0,0 +1,36 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# This configuration allows selecting what debug adapter debugging rpi_pico
```
```suggestion
# This configuration allows selecting what debug adapter debugging adafruit_macropad_rp2040
```

kartben replied:
dropped this section

ajf58 commented on boards/adafruit/macropad_rp2040/board.cmake:
```diff
@@ -0,0 +1,36 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# This configuration allows selecting what debug adapter debugging rpi_pico
+# by a command-line argument.
```
Non-blocking: I find the wording of this sentence difficult to understand.

kartben replied:
dropped

ajf58 commented on boards/adafruit/macropad_rp2040/board.yml:
```diff
@@ -0,0 +1,6 @@
+board:
+  name: adafruit_macropad_rp2040
```
Non-blocking: Adafruit boards seem to prefix the board name with the vendor name. This seems to be an exception, and perhaps for consistency this should stay.

kartben replied:
Not sure I understand, sorry

ajf58 replied:
If I look at most (well, a random sample) of boards, the `name` is the name of the board. It doesn't contain the vendor in it, e.g. `name: nucleo_f030r8` not `name: st_nucleo_f030r8`. I guess I'm unclear if the board is a "Adafruit MacroPad RP2040" by Adafruit, or "MacroPad RP2040". It's not really significant.

kartben replied:
ya Adafruit and a few other vendors are an exception so I just aimed at being consistent with other boards from them. I don't think there is a clear rule except that it tries to get as close as possible to the commercial name of the products (see "**Adafruit** MacroPad RP2040" [here](https://www.adafruit.com/product/5128), "NUCLEO-F030R8" [here](https://www.st.com/en/evaluation-tools/nucleo-f030r8.html). Also note how it's `rpi_pico` and `rpi_pico2` (although `rasperrypi` is the official vendor prefix...) :) 
IMO the name should always be exempt of the vendor but one should be able to append vendor name if they want to (*or* if it's necessary to remove ambiguities).

ajf58 reviewed (changes requested):
A couple of copy-paste issues I think

ajf58 commented on boards/adafruit/macropad_rp2040/adafruit_macropad_rp2040.dts:
```diff
@@ -0,0 +1,243 @@
+
+		encoder_button: encoder_button {
+			gpios = <&gpio0 0 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>;
+			zephyr,code = <INPUT_KEY_A>;
```
For more options from the upstream docs:
```
The BOOT button is available in CircuitPython at board.ENCODER_SWITCH or board.BUTTON. It is available in Arduino at PIN_SWITCH. Pressing the button grounds the pin.
```

kartben replied:
> Nit: Why `A` here ?
> 
> I would personally prefer ENTER, because it feels more "natural" for such a rotary+click encoder
> 
> ```suggestion
> 			zephyr,code = <INPUT_KEY_ENTER>;
> ```

No particular reason, ENTER sounds good :)

kartben replied:
switched to INPUT_KEY_ENTER

kartben commented on boards/adafruit/macropad_rp2040/Kconfig.adafruit_macropad_rp2040:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_ADAFRUIT_MACROPAD_RP2040
+	bool "Adafruit MacroPad RP2040"
```
Ouch. So much for trying to get Cursor to help me with what should have been the boring / hard to mess up bits :|

kartben commented on boards/adafruit/macropad_rp2040/adafruit_macropad_rp2040.dts:
```diff
@@ -0,0 +1,243 @@
+	usb_cdc_acm_uart: cdc_acm_uart0 {
+		compatible = "zephyr,cdc-acm-uart";
+		label = "Zephyr USB CDC-ACM";
+	};
```
dropped

kartben commented:
Addressed all feedback, hopefully.

Also tried running twister tests for the board which I should have thought of doing earlier:

```
/scripts/twister -p adafruit_macropad_rp2040
ZEPHYR_BASE unset, using "/home/kartben/zephyrproject/zephyr"
INFO    - Using Ninja..
INFO    - Zephyr version: v4.0.0-4900-g593c1045b1f2
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/kartben/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 16
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:  760/ 760  100%  built (not run):  510, filtered: 2768, failed:    0, error:    0error:    0
INFO    - 3491 test scenarios (3279 configurations) selected, 2768 configurations filtered (2519 by static filter, 249 at runtime).
INFO    - 1 of 511 executed test configurations passed (0.20%), 510 built (not run), 0 failed, 0 errored, with no warnings in 1234.75 seconds.
INFO    - 1 of 1 executed test cases passed (100.00%) on 1 out of total 934 platforms (0.11%).
INFO    - 4545 selected test cases not executed: 4545 not run (built only).
INFO    - 1 test configurations executed on platforms, 510 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/kartben/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/kartben/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/kartben/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

titouanc reviewed (approved):
As far as I'm concerned, LGTM

ajf58 commented on boards/adafruit/macropad_rp2040/board.cmake:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+board_runner_args(uf2 "--board-id=RPI-RP2")
```
It feels a little extreme to remove all the other things, but I guess this is something you've tested, whereas not the others? I imagine the lack of OpenOCD config will have reduced the amount of test coverage quite a bit though, no?

kartben replied:
![PXL_20250206_153010249.MP~3.jpg](https://github.com/user-attachments/assets/4dce64f2-fd8f-4c5b-bcb1-ab67babcc0a4)

This is how accessible SWDIO and SWCLK are on that board, so I think it's safe to stick to UF2 only, wdyt? :)
And several other rp2040 boards do the same, apparently. 
https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/adafruit/qt_py_rp2040/board.cmake
https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/seeed/xiao_rp2040/board.cmake

ajf58 reviewed (approved):
LGTM. All my concerns have been fixed, and comments are fine.
